### DISCUSSION TITLE
Question to LoRa E32 Config Mode / Transmission Mode | UART

### DISCUSSION THREAD

**Discussion by Jazzkeller** [micropython/micropython#11212] [RP2040 / Pico] [answered]
Hi,

I'm new to LoRa and experimenting with the E32-433T30D module. When I set M0/M1 high and run the following code, I got output via UART.

```
Initialization: {} Success
Retrieve configuration: {} Success
----------------------------------------
HEAD : 0b11000000 192

AddH : 1
AddL : 2
Chan : 23  ->  433

SpeedParityBit    : 0b0  ->  8N1 (Default)
SpeedUARTDatte : 0b11  ->  9600bps (default)
SpeedAirDataRate  : 0b10  ->  2.4kbps (default)
OptionTrans       : 0b0  ->  Transparent transmission (default)
OptionPullup      : 0b1  ->  TXD, RXD, AUX are push-pulls/pull-ups (default)
OptionWakeup      : 0b0  ->  250ms (default)
OptionFEC         : 0b1  ->  Turn on Forward Error Correction Switch (Default)
OptionPower       : 0b11  ->  21dBm
----------------------------------------
```

python:
```
from lora_e32 import LoRaE32, print_configuration, Configuration
from lora_e32_operation_constant import ResponseStatusCode
from machine import UART

uart0 = UART(1,9600)

lora = LoRaE32('433T30D', uart0)

code = lora.begin()
print("Initialization: {}", ResponseStatusCode.get_description(code))

code, configuration = lora.get_configuration()

print("Retrieve configuration: {}", ResponseStatusCode.get_description(code))

print_configuration(configuration)
```

When I set M0/M1 to GND and run the transmit data script, something with UART is wrong, I got the following output:
```
LoRa-E5 detected

Data: 
```

python:
```
from machine import UART, Pin
from utime import sleep_ms
from sys import exit

uart1 = UART(1, 9600)
join_EUI = None 
device_EUI = None

### Function Definitions

def receive_uart():
    rxData=bytes()
    while uart1.any()>0:
        rxData += uart1.read(1)
        sleep_ms(2)
    return rxData.decode('utf-8')

def send_AT(command):
    buffer = 'AT' + command + '\r\n'
    uart1.write(buffer)
    sleep_ms(300)

def test_uart_connection():
    send_AT('') # empty at command will query status
    data = receive_uart()
    if data == '+AT: OK\r\n' : print('LoRa radio is ready\n')
    else:
        print('LoRa-E5 detected\n')
        print('Data: '+data)
        exit()

def get_eui_from_radio():
    '''Reads both the DeviceEUI and JoinEUI from the device'''
    send_AT('+ID=DevEui')
    data = receive_uart()
    device_EUI = data.split()[2]

    send_AT('+ID=AppEui')
    data = receive_uart()
    join_EUI = data.split()[2]

    print(f'JoinEUI: {join_EUI}\n DevEUI: {device_EUI}')
   
test_uart_connection()

get_eui_from_radio()
```
So, seems to me that there's nothing coming back from the module via UART. Any idea what I'm doing wrong? I've done research in Google, but I couldn't find any solutions. Any idea what I'm doing wrong?

Btw: I'm a licensed Ham Radio Operator. I checked on 433 that the module is sending data with the script.

**Comment by peterhinch** [comment]
Your `receive_uart` function will quit with nothing if called before a complete character has been received. The timing for subsequent characters is questionable.

Consider using `uasrt.readline()` or `uart.read()`: in both cases a timeout can be specified.

**Reply by Jazzkeller** [reply]
Ok, but I got uart.read(1) in the receive_uart function. Isn't that sufficient?

**Reply by peterhinch** [reply]
If you call the function before a complete character has arrived, `uart.read()` never gets to run because `.any()` will return 0.

Writing a one-character at a time read routine is a respectable way to handle a UART, but it is a task which requires some experience and knowledge of UARTs. There are bear traps around timing. If your incoming data is newline-terminated, I strongly recommend the `readline()` method which is proven. A UART timeout can handle error conditions.

**Reply by mcauser** [reply]
uart.readline() works well up until you foolishly send yourself a line longer than the RX buffer length. That one got me a few times. 🤦

**Comment by mcauser** [comment]
I'm confused...
Can you provide any links to the libraries you're using?

The EByte E32-433T30D: https://www.ebyte.com/en/product-view-news.aspx?id=108
The Seeed LoRa E5: https://www.seeedstudio.com/LoRa-E5-Wireless-Module-p-4745.html

The E32 is a non-LoRaWAN transceiver (plain old LoRa packet radio) and the E5 includes the LoRaWAN stack.
Both are controlled via UART. Very different boards. Very different configuration interface.

The E32 is configured with a simple 5-bytes protocol when M0 and M1 are HIGH.
The E5 is configured with AT commands (using the stock firmware assuming you haven't made your own with STM32CubeMX)

Sounds like your mixing code for the two. If your code is dealing with EUI's, it's LoRaWAN.

When your E32's M0+M1 are LOW, it's in TX/RX mode. Anything you send over UART will be blasted out over the air.
If you have a 2nd E32, and it's configured to use the same channel and data rates, uart.write() on the first module will show up in uart.read() on the second one.

The transmission mode feature of the E32 lets you give devices addresses and when you TX data, you can send it to specific addresses.
All E32's in the are will RX the data, but if they aren't listening to that address will ignore the data. There's also a broadcast address to send and receive on, for when you want all devices to interact.

I've got both E32s and E5s, and unfortunately, I've been unable to get them to communicate (when you put the E5 in packet radio mode).
E32->E32 works great, E5->LoRaWAN works great.

Also, if you are buying more E32s, they recently upgraded the radio on the latest E32 models. More efficient and greater range, but incompatible with previous gen E32s. I think it was SX127x vs SX126x or something.

**Reply by Jazzkeller** [reply]
Ah ok got it, thanks! No LoRaWAN for the E32. I'm building 2 devices, peer to peer, which can send and receive data. Would that work with the E32 modules?

I connected the M0/M1 to GPIO Pins, I'm now able to send data or switch to config mode.

**Reply by Jazzkeller** [reply]
I'm using this library atm: https://github.com/xreef/EByte_LoRa_E32_micropython_library

**Reply by mcauser** [reply, marked as answer]
That's the use case for an E32 module. It replaces a wire connecting two devices with the magic of RF.
If you configure the two E32 modules the same, data sent to one comes out the other side.

If you have more than 2x E32s, that's where transmission modes come in. You can send/receive from/to specific devices using addresses.

Yeah, you can't leave M0/M1 floating. They need to be on GPIOs so you can toggle them to enter sleep (config) mode and normal operations.

**Reply by Jazzkeller** [reply]
Thanks! Appreciate your help. I made progress and I'm able to rcv/tx data!
